wasm/validator/src/visitor.rs

```diff
@@ -0,0 +1,1029 @@
+        validate_register_trigger(Register<Trigger<FilterBox, Executable>>),
+        validate_register_permission_token(Register<PermissionTokenDefinition>),
+
+        // Validate UnregisterBox
+        validate_unregister_peer(Unregister<Peer>),
+        validate_unregister_domain(Unregister<Domain>),
+        validate_unregister_account(Unregister<Account>),
+        validate_unregister_asset_definition(Unregister<AssetDefinition>),
+        validate_unregister_asset(Unregister<Asset>),
+        validate_unregister_role(Unregister<Role>),
+        validate_unregister_trigger(Unregister<Trigger<FilterBox, Executable>>),
+
+        // Validate MintBox
+        validate_mint_asset(Mint<Asset, NumericValue>),
+        validate_mint_account_public_key(Mint<Account, PublicKey>),
+        validate_mint_account_signature_check_condition(Mint<Account, SignatureCheckCondition>),
+        validate_mint_trigger_repetitions(Mint<Trigger<FilterBox, Executable>, u32>),
+
+        // Validate BurnBox
+        validate_burn_account_public_key(Burn<Account, PublicKey>),
+        validate_burn_asset(Burn<Asset, NumericValue>),
+
+        // Validate TransferBox
+        validate_transfer_asset_definition(Transfer<Account, AssetDefinition, Account>),
+        validate_transfer_asset(Transfer<Asset, NumericValue, Asset>),
+
+        // Validate SetKeyValueBox
+        validate_set_domain_key_value(SetKeyValue<Domain, Name, Value>),
+        validate_set_account_key_value(SetKeyValue<Account, Name, Value>),
+        validate_set_asset_definition_key_value(SetKeyValue<AssetDefinition, Name, Value>),
+        validate_set_asset_key_value(SetKeyValue<Asset, Name, Value>),
+
+        // Validate RemoveKeyValueBox
+        validate_remove_domain_key_value(RemoveKeyValue<Domain, Name>),
+        validate_remove_account_key_value(RemoveKeyValue<Account, Name>),
+        validate_remove_asset_definition_key_value(RemoveKeyValue<AssetDefinition, Name>),
+        validate_remove_asset_key_value(RemoveKeyValue<Asset, Name>),
+
+        // Validate GrantBox
+        validate_grant_account_permission(Grant<Account, PermissionToken>),
+        validate_grant_account_role(Grant<Account, RoleId>),
+
+        // Validate RevokeBox
+        validate_revoke_account_permission(Revoke<Account, PermissionToken>),
+        validate_revoke_account_role(Revoke<Account, RoleId>),
+
+        // Validate UpgradeBox
+        validate_upgrade_validator(Upgrade<Validator>),
+    }
+}
+
+/// Default validation for [`SignedTransaction`].
+///
+/// Performs execution if transaction contains [`Executable::Instructions`]
+/// and does nothing if [`Executable::Wasm`].
+///
+/// Execution is done to properly validate dependent instructions.
+pub fn validate_transaction<V: Validate + ?Sized>(
+    validator: &mut V,
+    authority: &AccountId,
+    transaction: &SignedTransaction,
+) -> Verdict {
+    let verdict = validator.validate_transaction_signatures(authority, transaction.signatures());
```

Comment:
> With that it turns out that we have `SignatureCheckCondition` and another source of truth insinde validator

it's not two sources of truth because this doesn't override `SignatureCheckCondition`, it adds additional validation to it. They are orthogonal

Reply:
Hm, I'm not sure right now. Can we have @appetrosyan attention here?

Reply:
I'm not sure of potential use case for this.
But it seems like this checks will happen at different stages and their failure cause different effects. 
In case of `SignatureCheckCondition` transaction event won't be added to the `Queue` (dropped essentially).
And in case of `validate_transaction_signatures` transaction will be added to rejected transactions.

Reply:
IMO this is a visitor and it visits every node in the `Transaction` syntax tree.

Reply:
Okay, fair enough, let's keep this